repos/system_upgrade/el8toel9/actors/checkvdo/libraries/checkvdo.py

```diff
@@ -0,0 +1,180 @@
+    exit_code = result['exit_code']
+    # 255: Not a vdo device
+    #   0: A post-migration vdo device
+    #   1: A pre-migration vdo device
+    if (exit_code != 255) and (exit_code != 0) and (exit_code != 1):
+        raise exceptions.StopActorExecutionError(
+                'failed to check device {0}; exit code: {1}'.format(device, exit_code))
+    return exit_code == 0
+
+
+def _get_migration_failed_vdo_devices():
+    """ Collect all vdo devices that have been migrated for lvm-based management. """
+    lsblk_info = _get_migration_failed_lsblk_info()
+
+    # Exclude ROM and VDO devices.
+    # Any vdo devices appearing here are unmigrated (lsblk does not identify
+    # lvm-managed vdos as vdos); we could leave them in and rely on the
+    # post-migration device check, but there's no reason to do so.
+    devices = set([device for (device, info) in lsblk_info.items()
+                          if info['type'] not in ['rom', 'vdo']])
+
+    # VDO devices that have been created by or migrated to lvm-based management
+    # are not identified by blkid.
+    # Remove any devices that blkid can identify.
+    devices -= set(_get_migration_failed_canonical_blkid_results())
+
+    return [x for x in devices if _is_post_migration_vdo_device(x)]
+
+
+# Check for vdo devices that must be migrated to lvm management.
+def _check_for_unmigrated_vdo_devices():
+    # Find VDO devices that have not been migrated to lvm-based management.
+    unmigrated_vdo_devices = _get_unmigrated_vdo_devices()
+
+    if len(unmigrated_vdo_devices) == 0:
+        # Generate report indicating there are no vdo devices to migrate.
+        reporting.create_report([
+            reporting.Title('VDO devices migration to lvm-based management'),
+            reporting.Summary('VDO devices that require migration: None'),
+            reporting.Severity(reporting.Severity.LOW),
+            reporting.Tags([reporting.Tags.SERVICES, reporting.Tags.DRIVERS])
+        ])
+    else:
+        devices = [device for device in unmigrated_vdo_devices]
+        reporting.create_report([
+            reporting.Title('VDO devices migration to lvm-based management'),
+            reporting.Summary('VDO devices that require migration: {0}'.format(', '.join(devices))),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.SERVICES, reporting.Tags.DRIVERS]),
+            reporting.Remediation(hint = 'Perform VDO to LVM migration on the VDO devices.'),
+            reporting.Flags([reporting.Flags.INHIBITOR])
+        ])
+
+
+def _check_for_migration_failed_vdo_devices():
+    # Find VDO devices that did not complete migration to lvm-based management.
+    # This could result from system failures during the migration process.
+    migration_failed_vdo_devices = _get_migration_failed_vdo_devices()
+    if len(migration_failed_vdo_devices) > 0:
+        devices = [device for device in migration_failed_vdo_devices]
```

Comment:
```suggestion
```
1. `_get_migration_failed_vdo_devices` already returns a list
2. It's not necessary even if it wouldn't be a list as `str.join` can handle any iterable which a set is.
3. If you must use just `devices = list(devices)` which will give you the same result and is more terse. No need for a list comprehension.

Reply:
Addressed in commit 49eb4a6.